### DISCUSSION TITLE
Update default entity_id when name of a device changes outside of HA

### DISCUSSION THREAD

**Discussion by marcelveldt** [home-assistant/architecture#741] [General]
Speaking of "streamlining" and "make things easier" I think this is one that confuses a lot of users: the enity_id not matching the actual entity name.

**The issue**
Home Assistant creates the entity_id based on the entity name. In 99% of the cases the entity name is derived from the device name. Although entities actually have a much better, stable identifier (unique_id) this is not in visible/usable in the Home Assistant UI, automations and scripts.

In a happy world the device name is changed from within HA and the UI offers to also update the entity names and entity_id's.
In most cases however the devices (and child entities) are retrieved from some hub or cloud service which allows updating the name there, outside of HA.

For example in case of Hue I add a new light to my setup.
Hue names that new light by default "Hue Ambience Spot" or something other model related.
Home Assistant gets informed by the newly added light and bravely creates a new device and entities for this fancy lightbulb, using that same name. The entity_id for the new light will be "light.hue_ambience_spot" or when I'm adding more I'll end up with "light.hue_ambience_spot_2" etc.

Now the Hue app offers me to give a new personal name to the light, for example I name it "Desk light office".
Home Assistant gets informed by the device name update and this is properly reflected in the UI. The device name is now "Desk light office" and so is the light entity. Except for the entity_id, which stayed at "light.hue_ambience_spot" which feels very wrong, especially since the fact that I just created this new light.

I've used Hue here as an example but any integration suffers from this issue where you can set device/entity names outside of HA.

Future wise you could say that the entity_id should be phased out in favor of the stable unique_id (and just make selectors everywhere to just select the pretty name) but that is maybe a bridge too far (for now).

For now I'm proposing a smart way to detect name changes at the entity registry level and update entity_id's based on that:

- Integration sets new name attribute on entity and schedules update
- Entity registry performs checks: 
   -- "did the name change?"
   -- "is the entity name not user overridden?"
   -- "does the slugified previous entity name exactly resolve to the entity_id?"
- If checks pass, it means the entity_id was auto created from the entity name before and not changed by the user.
- Update the entity_id to a slugified version of the new entity name

Now comes the hard part: This entity could already been used in scripts/automations etc. How to handle that ?

A few ideas:
1) Only auto update the entity_id if the old id has not been used. Log otherwise.
2) Always do the rename action, create some sort of very visual alert that the user should rename the id in scripts etc.
3) Try to rename the entity_id change in scripts etc. although that will be practically impossible.

**Alternative solution**
Do not auto create the entity_id based on the friendly name but use a (more) stable identifier as the hint/default for the entity_id, such as the internal id within the integration (unique_id) or just derrive it from the device and model name instead of the customized friendly name.

That way the impression is gone that it should auto update UX wise.

**Comment by frenck** [comment]
It sounds to me like you are trying to re-introduce a method we have been driving away from.

I'm strongly against changing entity IDs automatically. The entity registry was part of stabilizing it. We should not touch entity IDs once registered.

Simply put: We do not know when and where they are used. We cannot update automations and cannot update scripts automatically. Additionally, they can be used and referenced outside of Home Assistant as well (e.g., Node-RED, AppDaemon, NetDaemon, or anything else using our APIs).

> Do not auto create the entity_id based on the friendly name but just create some random identifier or use the unique_id

Integrations can already do that. Please note, a entity ID should not be random. We have a unique identifier. The entity ID is a human consumable (and customizable) identifier.

**Reply by marcelveldt** [reply]
Well, I'm only observing the fact that people get a bit confused by the entity_id's from a UX perspective. I've ran into it myself and seen multiple reports and I think it can be improved. My solution is just a proposal to discuss about or have feedback on (get the discussion started).

I hear you have a strong opinion about not changing them, I have a strong opinion that you should not confuse users. In line of the current motto of "streamlining things" and making things more accessible for non tech savvy users, this is one I wanted to write down to have a discussion about.

The solution may as well be a common agreement to use a stable human consumable identifier that is not attached to the name.
My observation is that most integrations (and yes, I'm also to blame for that) seem to prefer the user customized friendly name for the device and that causes the confusion. If for example the (default) entity_id is based on e.g. the manufacturer/device model the issue is already partly gone.

**Reply by marcelveldt** [reply]
> Integrations can already do that. Please note, a entity ID should not be random. We have a unique identifier. The entity ID is a human consumable (and customizable) identifier.

And is this enforced in accepting integrations and platforms to guard for a unified experience ?
is this well pointed out in the developer documentation ? I got triggered by your sentence "Integrations can already do that" and now found out that this even exists :-)

**Reply by frenck** [reply]
> I hear you have a strong opinion about not changing them

Yeah, because we are not UX only focussed system. Streamlining != UX/UI only.

> The solution may as well be a common agreement to use a stable human consumable identifier that is not attached to the name.

Agreed, which is already possible to implement as an integration sees fit.
Not all integrations have a better "suggestion" than a name though, so this can be implemented as fit on a integration by integration base.

The downside of that approach is though: Entity IDs are no longer predictable / as expected maybe. As behavior across integrations will change.

**Reply by marcelveldt** [reply]
> Agreed, which is already possible to implement as an integration sees fit.
>Not all integrations have a better "suggestion" than a name though, so this can be implemented as fit on a integration by integration base.

So this could be something for the developer docs to add as best practice advice and adjust the top10 integrations according to that best practice advice.

> The downside of that approach is though: Entity IDs are no longer predictable / as expected maybe. As behavior across integrations will change.

Yeah, but that's becoming less of an issue where the entity_id is only used as "legacy" id for advanced users only ?
All UI screens will just present the user with a dropdown to select the entity.

**Comment by balloob** [comment]
So we're working on two different solutions to this problem:

 - Do not show the entity ID in the UI. Users in the interface should never have to deal with internal identifiers.
 - Entities are registered in the entity registry based on the unique ID of an entity. The entity registry entry has it's own primary key that will never change. We've started to accept these for things, like [the `target` field](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/config_validation.py#L995) of a service.

**Reply by frenck** [reply]
> Do not show the entity ID in the UI.

Yeah, I think we can hide it from almost all places in the UI, which I agree on.

That said, it is the primary driver for anything manual/advanced, so it should still be accessible in some way.

**Reply by marcelveldt** [reply]
I think this is the way forward and kind of what I said in my story above that ultimately a user should not be confronted with id's, unless it's an advanced user and/or advanced usecase.

That combined with a sane default for the entity_id when it's created is the solution and avoid a lot of confusion.
The target fields however are still based on entity_id or does it already accept the entity uuid from entity registry ?

**Reply by frenck** [reply]
> The target fields however are still based on entity_id or does it already accept the entity uuid from entity registry ?

There have been multiple PRs enabling that by @emontnemery. I don't believe the full scope has been completed at this point.

**Reply by marcelveldt** [reply]
That's great news to hear!
If we can settle on what we like as default for the entity_id "hint" I can create some PR's to adjust integrations
